featurebyte/schema/feature_job_setting_analysis.py

```diff
@@ -112,11 +112,33 @@ class FeatureJobSetting(FeatureByteBaseModel):
     Feature Job Setting
     """
 
-    frequency: int
-    job_time_modulo_frequency: int
```

Comment:
Will this break the UI? Wondering if it's possible to make this backward compatible for a smoother transition.

Reply:
There are list of items that we need to update the UI. This is not the only place that requires changes in UI. 
If we want to do that, we may need to keep both old & new parameters appear at the response, and at the same time, our migration should only keep the new feature job setting format. That means we have the `FeatureJobSetting` behave differently when it is at the response versus the format used in persistence. It requires more changes & updates.

Reply:
Yes, I was wondering whether it's possible to keep both old and new parameters in the response without too much work. Something like

```python
class FeatureJobSetting(FeatureByteBaseModel):
    """
    Feature Job Setting
    """

    period: int
    offset: int
    ...
    frequency: int = Field(exclude=True)
    job_time_modulo_frequency: int = Field(exclude=True)

    @root_validator(pre=True)
    @classmethod
    def _handle_backward_compatibility(cls, values: Dict[str, Any]) -> Dict[str, Any]:
        # set period / offset using frequency / job_time_modulo_frequency if provided
        ...
        # set frequency / job_time_modulo_frequency using period / offset if provided
        ...
```

and backend will still always use the new parameters `period` and `offset` and ignore the old parameters. Will this work and eliminate UI breakage?

Reply:
If the `exclude` is set to `True`, `feature_job_setting.dict()` will exclude those parameters too. That means response will not have these parameters set.

Reply:
Inside the fastapi response preparation:
https://github.com/tiangolo/fastapi/blob/0.96.0/fastapi/routing.py#L79-L84

We also use `dict` when we serialize the model into dictionary format before writing to mongo persistent.

Reply:
Unless we want to overwrite pydantic `BaseModel.dict` behaviour so that it can behave differently for response preparation & persistent input preparation.

Reply:
I see. How about using a different schema for read response? Then it is decoupled from what is stored in persistent.

Reply:
Ya, i think this is possible for those response where the feature job setting is not deeply nested inside the schema (for example feature's graph).

Reply:
I think it would be nice if we can avoid breaking UI if the effort is not too big, but otherwise I'm also fine with the current approach.

Reply:
Only the feature's graph level that part is difficult (skipping this deeply nested fjs inside the graph should be fine). Others should be Ok. Will update this PR & tests it on the featurebyte-app first. 🙏